### ISSUE TITLE
var and type apparent for nullable types

### ISSUE THREAD

**Issue by brinko99** [dotnet/roslyn#73312]
Regarding `csharp_style_var_when_type_is_apparent`, we'd like to suggest an enhancement to the current logic or an additional rule disallowing `var` for nullable types.

In our opinion, nullable types should not be considered 'apparent' when in the nullable context.

For example below, the type is apparent, but not necessarily the fact that it's nullable. `var` hides the fact that the variable is nullable, making the code less clear as `applicationAssembly` is later used.

`Assembly? applicationAssembly = Assembly.GetEntryAssembly();`

**Comment by brinko99** [comment]
Thanks for your response. Can you clarify? I don't understand your comment "prevent 'var' from ever being used".

I'm imagining "apparent" logic that could act as such for different signatures:

`Assembly GetEntryAssembly()' -- Type is apparent, use `var`
`Assembly? GetEntryAssembly()' -- Type is not apparent, use explicit type

Per your second point, as compatibility is a concern, I suggest a new style option that requires an explicit type used when the type is nullable in a nullable context. This could be defaulted to off to ensure compatibility.

**Comment by CyrusNajmabadi** [comment]
> Can you clarify? I don't understand your comment "prevent 'var' from ever being used".

in all place where you use 'var' with a reference type, the type is `T?` from teh perspective of the language/compiler.

You said:

> In our opinion, nullable types should not be considered 'apparent' when in the nullable context.

That would mean that if you turned on `csharp_style_var_when_type_is_apparent` we would not ever suggest 'var' for a reference type in nullable contexts, which would be massively bad :)

**Comment by CyrusNajmabadi** [comment]
> Per your second point, as compatibility is a concern, I suggest a new style option that requires an explicit type used when the type is nullable in a nullable context

In hte past we've rejected this idea as every single developer has a different view on this topic.  We went with the majority view when designing the option and have stuck with that to keep broad consistency across the ecosystem.  We do not think knobs/options helps here.

**Comment by pjannesen** [comment]
> reference type



> > var hides the fact that the variable is nullable
> 
> This would prevent 'var' from ever being used. As far as the language is concerned, 'var' is always nullable for reference types.
> 
> We also cannot change this as that will cause build breaks for anyone upgrading.

For a long time, 'var' has been preferred to explicit typing. Because it takes less typing, more readable and easier to maintain.

That 'var' doesn't inherit the nullable constraints of reference types, makes 'var' virtually useless with '<Nullable>enabled</Nullable>'.

I understand that this cannot be solved by default. Because it will break the existing builds.

Is it possible to get an additional option? For example, '<Nullable>enabled+var</Nullable>' so that 'var' follows nullable constraints on reference types?

**Comment by CyrusNajmabadi** [comment]
> That 'var' doesn't inherit the nullable constraints of reference types, makes 'var' virtually useless with 'enabled'.

We use var with nullable-enabled in Roslyn.  It works great.  What issue are you running into?

**Comment by pjannesen** [comment]
You're right. I was misled by the var reference type being nullable. This is what I have found so far is:

Apparently there is another flag beside of the type definition. This flag defines if de value is 'may be null' or 'is not null'. Type definition is irrelevant if a value 'may be null'/'is not null'.

- type definition ( ? or no ? only important for parameters, returns, properties. Not relevant inside a method.
- Flow analysis determined whether that value 'may be null' or 'is not null'.

Examples:
- The variable is assign with reference or tested != null. Then is the variable 'is not null'
Examples:
```
string? x=_function() /* function return is string? */
if (x != null) {
	/* x is not null is this block */
}
```

- If a variable is tested for null outside the block. the value become 'may be null'. This is correct, but confusing. (This one was de reason why I was googling what is going on)
Example:
```
string v = "1"
if (v != null) { Console.WriteLine(v) }
/* v may by null besides the type definition is not null! */
```
